packages/components/src/input-control/reducer/state.ts

```diff
@@ -19,7 +19,10 @@ export interface InputState {
 	value?: string;
 }
 
-export type StateReducer = Reducer< InputState, InputAction >;
+export type StateReducer< SpecializedAction = Action > = Reducer<
+	InputState,
+	InputAction | SpecializedAction
+>;
```

Comment:
In the current state of this PR, the `StateReducer` type would accept _any_ action-like object (this can be verified by inspecting [`action.type`](https://github.com/WordPress/gutenberg/blob/3b54cff387981613a3074ff3338c9bbbe438566f/packages/components/src/input-control/reducer/reducer.ts#L57) and seeing what possible values are attributed to it by the TypeScript engine).

Instead, what we'd like to achieve is:
- When `SpecializedAction` is not defined, `StateReducer` should only allow actions from the `InputActions` type
- When `SpecializedAction` is defined, `StateReducer` should only allow actions from either the `InputActions` type, or the `SpecializedAction` type (although we need to make sure that `SpecializedAction` is a valid `Action`).

I played around with code a bit and I think this should be a good solution:

```suggestion
export type StateReducer< SpecializedAction = {} > = Reducer<
	InputState,
	SpecializedAction extends Action
		? InputAction | SpecializedAction
		: InputAction
>;
```

Reply:
Very neat Marco! Applied in 30d7c31. I don't recall having seen such logic in typing. TIL 🙌

Reply:
Thank you! I learnt this technique by reading the definition of the `WordPressComponentProps` type 
https://github.com/WordPress/gutenberg/blob/33d84b036592a5bf31af05b7710f3b2b14163dc4/packages/components/src/ui/context/wordpress-component.ts#L19-L24